pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -630,6 +682,26 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *v1.Pod, ru
 		}(container)
 	}
 	wg.Wait()
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarLifecycle) && len(sidecars) != 0 {
+		gracePeriodOverride = gracePeriodRemaining(pod, gracePeriodOverride, timeContainerKillStarted)
+
+		wg.Add(len(sidecars))
+		for _, container := range sidecars {
+			go func(container *kubecontainer.Container) {
+				defer utilruntime.HandleCrash()
+				defer wg.Done()
+
+				killContainerResult := kubecontainer.NewSyncResult(kubecontainer.KillContainer, container.Name)
+				if err := m.killContainer(pod, container.ID, container.Name, "", gracePeriodOverride); err != nil {
```

Comment:
I'm 95% sure I'm missing something :) But are we not going to run the `preStopHook` twice for side car containers? Because we already called it, and now we are killing `killContainer`, which also tries to run the pre-stop lifecycle hooks.

Reply:
no that's correct it would get called twice, pre-stop hooks already have no guarantee that they won't get called multiple times so they're meant to be idempotent, we discussed this in the design phase and agreed that as long as we call it out it should be ok. From the docs:
> Hook delivery is intended to be at least once, which means that a hook may be called multiple times for any given event, such as for PostStart or PreStop. It is up to the hook implementation to handle this correctly.

Reply:
Ahh gotcha! Today I learned :) thanks!